`azurerm_windows_function_app`, `azurerm_linux_function_app` - fix content share not being created automatically issue

xiaxyi opened the pull request:
Fix content share not being created automatically if `WEBSITE_CONTENTOVERVNET` sets to 1.

According to the official [guidance](https://docs.microsoft.com/en-us/azure/azure-functions/functions-app-settings#website_contentshare), _"There are scenarios where you must set the WEBSITE_CONTENTSHARE value to a predefined share, such as when you [use a secured storage account in a virtual network](https://docs.microsoft.com/en-us/azure/azure-functions/configure-networking-how-to#restrict-your-storage-account-to-a-virtual-network). In this case, you must set a unique share name for the main function app and the app for each deployment slot."_
If the property `WEBSITE_CONTENTOVERVNET` set to 1 in `app_setting` while the users don't specify or specify a non-existed storage file share, they get the error message:

```
reading logs configuration for Windows Function App: (Site Name "xxx" / Resource Group "xxx"): web.AppsClient#GetDiagnosticLogsConfiguration: Failure responding to request: StatusCode=503 -- Original Error: autorest/azure: Service returned an error. Status=503 Code="ServiceUnavailable" Message="The operation timed out and could not be completed. Please retry the action or try again later." Details=[{"Message":"The operation timed out and could not be completed. Please retry the action or try again later."},{"Code":"ServiceUnavailable"},{"ErrorEntity":{"Code":"ServiceUnavailable","ExtendedCode":"51007","Message":"The operation timed out and could not be completed. Please retry the action or try again later.","MessageTemplate":"The operation timed out and could not be completed. Please retry the action or try again later."}}]
│
```

fix https://github.com/hashicorp/terraform-provider-azurerm/issues/17254
fix https://github.com/hashicorp/terraform-provider-azurerm/issues/14167
fix #19134

jackofallops commented on internal/services/appservice/helpers/service_plan.go:
```diff
@@ -137,7 +137,7 @@ func PlanTypeFromSku(input string) string {
 
 // ServicePlanInfoForApp returns the OS type and Service Plan SKU for a given App Service Resource
-func ServicePlanInfoForApp(ctx context.Context, metadata sdk.ResourceMetaData, id interface{}) (osType *string, planSku *string, err error) {
+func ServicePlanInfoForApp(ctx context.Context, metadata sdk.ResourceMetaData, id interface{}, serviceFarmId string) (osType *string, planSku *string, err error) {
```
This function is intended to retrieve the Service Plan ID from the App ID, it should not take the ID directly to circumvent this.

jackofallops commented on internal/services/appservice/linux_function_app_resource.go:
```diff
@@ -703,19 +767,21 @@ func (r LinuxFunctionAppResource) Update() sdk.ResourceFunc {
 			}
 
-			_, planSKU, err := helpers.ServicePlanInfoForApp(ctx, metadata, *id)
+			// Some service plan updates are allowed - see customiseDiff for exceptions
+			var serviceFarmId string
+			if metadata.ResourceData.HasChange("service_plan_id") {
+				serviceFarmId = state.ServicePlanId
+				existing.SiteProperties.ServerFarmID = utils.String(serviceFarmId)
+			}
+
+			_, planSKU, err := helpers.ServicePlanInfoForApp(ctx, metadata, *id, serviceFarmId)
```
This should be a call to the new service plan to get the SKU, and not repurpose the helper function.

jackofallops commented on internal/services/appservice/windows_function_app_resource.go:
```diff
@@ -536,6 +547,22 @@ func (r WindowsFunctionAppResource) Read() sdk.ResourceFunc {
 				return fmt.Errorf("reading Windows %s: %+v", id, err)
 			}
 
+			var planSku *string
+			if functionApp.ServerFarmID != nil {
+				servicePlanId, err := parse.ServicePlanID(*functionApp.ServerFarmID)
+				if err != nil {
+					return err
+				}
+				servicePlan, err := servicePlanClient.Get(ctx, servicePlanId.ResourceGroup, servicePlanId.ServerfarmName)
+				if err != nil {
+					return fmt.Errorf("reading %s: %+v", servicePlanId, err)
+				}
+				if sku := servicePlan.Sku; sku != nil && sku.Name != nil {
+					planSku = sku.Name
+				}
+			}
+			sendContentSettings := helpers.PlanIsElastic(planSku)
+
```
As above, this functionality shouldn't be present in the Read.

xiaxyi replied:
We need to get the service plan id to determine if the function app is elastic or not, then to check if the website_content_share property.

jackofallops commented on internal/services/appservice/windows_function_app_resource.go:
```diff
@@ -546,6 +573,43 @@ func (r WindowsFunctionAppResource) Read() sdk.ResourceFunc {
 				return fmt.Errorf("reading App Settings for Windows %s: %+v", id, err)
 			}
 
+			var storageAccountName, storageFileShare string
+			if sendContentSettings || helpers.PlanIsConsumption(planSku) {
+				if appSettingsResp.Properties != nil {
+					if appSettingsResp.Properties["WEBSITE_CONTENTAZUREFILECONNECTIONSTRING"] != nil {
+						saConnectionString := strings.Split(*appSettingsResp.Properties["WEBSITE_CONTENTAZUREFILECONNECTIONSTRING"], ";")
+						for _, part := range saConnectionString {
+							if strings.Contains(part, "AccountName") {
+								storageAccountName = strings.TrimPrefix(part, "AccountName=")
+								break
+							}
+						}
+					}
+					if appSettingsResp.Properties["WEBSITE_CONTENTSHARE"] != nil {
+						storageFileShare = *appSettingsResp.Properties["WEBSITE_CONTENTSHARE"]
+					}
+					saClient := metadata.Client.Storage
+					account, err := saClient.FindAccount(ctx, storageAccountName)
+					if err != nil {
+						return fmt.Errorf("retrieving Account %q for Share %q: %s", storageAccountName, storageFileShare, err)
+					}
+					if account == nil {
+						return fmt.Errorf("unable to locate Storage Account %s", storageAccountName)
+					}
+					saFileShareClient, err := metadata.Client.Storage.FileSharesClient(ctx, *account)
+					if err != nil {
+						return fmt.Errorf("building File Share Client: %s", err)
+					}
+					exists, err := saFileShareClient.Exists(ctx, account.ResourceGroup, storageAccountName, storageFileShare)
+					if err != nil {
+						return fmt.Errorf("checking for existence of Storage Share %q (Account %q) : %+v", storageFileShare, storageAccountName, err)
+					}
+					if exists == nil {
+						return fmt.Errorf("the storage file share %s for function app %s does not exists, please specify an existing one", storageFileShare, id.ID())
+					}
+				}
+			}
+
```
As above, this shouldn't be in the Read.

xiaxyi replied:
Thanks for the review, this is to address the timeout issue. if the user doesn't specify a predefined file share and also set the WEBSITE_CONTENTOVERVNET in the app_setting, the provider will timeout by trying to access a non-existence file share. I can add the 4.0 condition to this check block.

tombuildsstuff replied:
agreed - this shouldn't be in the Read function - we don't retrieve the file share in the Read, so this is unnecessary

xiaxyi replied:
Thanks @tombuildsstuff for the comments. If we don't check the storage file share, shall we just let the read function timeout?

tombuildsstuff replied:
@xiaxyi why will the read function timeout? it's calling the ARM API and interacting with the Storage API at all?

xiaxyi replied:
The function app is trying to connect to the file share specified in the `[WEBSITE_CONTENTSHARE](https://learn.microsoft.com/en-us/azure/azure-functions/functions-app-settings)`, Generally, the file share is created by function app automatically, however, the it cannot be created if user specify the key `WEBSITE_CONTENTOVERVNET` in `app_setting, it requires the user to create a valid file share in advance and put the value in `WEBSITE_CONTENTSHARE`. We didn't add such notes to user in the current provider, so the function(Read function) times out.

If we don't check the file share in read function, then what's left is to add notes in docs and let the read function timeout.

jackofallops commented on internal/services/appservice/windows_function_app_resource.go:
```diff
@@ -702,7 +766,13 @@ func (r WindowsFunctionAppResource) Update() sdk.ResourceFunc {
 			}
 
-			_, planSKU, err := helpers.ServicePlanInfoForApp(ctx, metadata, *id)
+			var serviceFarmId string
+			if metadata.ResourceData.HasChange("service_plan_id") {
+				serviceFarmId = state.ServicePlanId
+				existing.SiteProperties.ServerFarmID = utils.String(serviceFarmId)
+			}
+
+			_, planSKU, err := helpers.ServicePlanInfoForApp(ctx, metadata, *id, serviceFarmId)
```
As above, the helper should not be repurposed, and the SKU should be read from the new service plan ID. A new helper for that would be acceptable.

jackofallops commented on internal/services/appservice/linux_function_app_resource.go:
```diff
@@ -547,6 +575,43 @@ func (r LinuxFunctionAppResource) Read() sdk.ResourceFunc {
 				return fmt.Errorf("reading App Settings for Linux %s: %+v", id, err)
 			}
 
+			var storageAccountName, storageFileShare string
+			if sendContentSettings || helpers.PlanIsConsumption(planSku) {
+				if appSettingsResp.Properties != nil {
+					if appSettingsResp.Properties["WEBSITE_CONTENTAZUREFILECONNECTIONSTRING"] != nil {
+						saConnectionString := strings.Split(*appSettingsResp.Properties["WEBSITE_CONTENTAZUREFILECONNECTIONSTRING"], ";")
+						for _, part := range saConnectionString {
+							if strings.Contains(part, "AccountName") {
+								storageAccountName = strings.TrimPrefix(part, "AccountName=")
+								break
+							}
+						}
+					}
+					if appSettingsResp.Properties["WEBSITE_CONTENTSHARE"] != nil {
+						storageFileShare = *appSettingsResp.Properties["WEBSITE_CONTENTSHARE"]
+					}
+					saClient := metadata.Client.Storage
+					account, err := saClient.FindAccount(ctx, storageAccountName)
+					if err != nil {
+						return fmt.Errorf("retrieving Account %q for Share %q: %s", storageAccountName, storageFileShare, err)
+					}
+					if account == nil {
+						return fmt.Errorf("unable to locate Storage Account %s", storageAccountName)
+					}
+					saFileShareClient, err := metadata.Client.Storage.FileSharesClient(ctx, *account)
+					if err != nil {
+						return fmt.Errorf("building File Share Client: %s", err)
+					}
+					exists, err := saFileShareClient.Exists(ctx, account.ResourceGroup, storageAccountName, storageFileShare)
+					if err != nil {
+						return fmt.Errorf("checking for existence of Storage Share %q (Account %q) : %+v", storageFileShare, storageAccountName, err)
+					}
+					if exists == nil {
+						return fmt.Errorf("the storage file share %s for function app %s does not exists, please specify an existing one", storageFileShare, id.ID())
+					}
+				}
+			}
+
```
As above, none of this should be in a read function, we should be setting received data back into state only, not performing configuration checks on external resources that we're being told about.

xiaxyi replied:
the same to the reply in the windows function app

jackofallops reviewed (changes requested):
Hi @xiaxyi - Thanks for this PR.

I've left some comments inline below that will need to be addressed before we can continue review.  I think this PR should probably have been split into smaller specific PRs to address the disparate concerns as the changes do not fully match the description / title, which makes the changes difficult to review and will unfortunately slow the process. That said, if you can make the changes here we can get this moving.

Thanks!

xiaxyi commented:
> Hi @xiaxyi - Thanks for this PR.
> 
> I've left some comments inline below that will need to be addressed before we can continue review. I think this PR should probably have been split into smaller specific PRs to address the disparate concerns as the changes do not fully match the description / title, which makes the changes difficult to review and will unfortunately slow the process. That said, if you can make the changes here we can get this moving.
> 
> Thanks!

Thanks @jackofallops for the feedbacks. All the code changes in this pr is tend to resolve one issue which is the content share issue under elastic plan.

As I mentioned in the description, if user sets `WEBSITE_CONTENTOVERVNET ` without specifying a predefined storage file share, there will be error accessing that share. 

The file share check in read function is to ensure the predefined file share exists, otherwise, we throw error instead of waiting for the read function to time out. 

The service plan check is to determine whether the function is elastic plan.

xiaxyi commented:
@jackofallops @tombuildsstuff , updated th read function, removed the storage file share check.

jackofallops reviewed (changes requested):
Hi @xiaxyi - apologies for the long overdue re-review on this. I hope you don't mind, but I pushed some changes to simplify the updates and remove the need for re-processing some data.  Could you look to add a test to cover the case being addressed here, and I think we'll be good to merge once that's in and passing.

Thanks

xiaxyi commented:
Thanks @jackofallops for the update, I updated the code and added the test cases for both linux and windows.

feel free to let me know if there is anything else needed.

jackofallops reviewed (changes requested):
Hi @xiaxyi 
Thanks for adding the tests. I've re-pushed the changes that were lost during the merge main. The ignores in the test suggest this needs some more work, possibly in the functions I've called out in line below if you can take a look?

Thanks!

xiaxyi commented:
Thanks @jackofallops for the comments, I applied your change and pushed again, apologize for my mistakes last time. 

Regarding the acc test import steps, I think we still need to keep the ignore step for those app setting:
````
"app_settings.WEBSITE_CONTENTSHARE", "app_settings.WEBSITE_CONTENTAZUREFILECONNECTIONSTRING", "app_settings.%"
```

As I mentioned in my previous comments, service will try to add the `WEBSITE_CONTENTSHARE` and `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING` for the function app instance (in elastic or consumption plan) automatically if no `WEBSITE_CONTENTOVERVNET` is added by user, which means, we need to determine whether to set those kvp back to state by checking the user's TF config and that's why we need to ignore the changes when doing the import. So I'm not going to change the original unpackappsetting function:https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/appservice/linux_function_app_resource.go#L1090

jackofallops commented:
> Thanks @jackofallops for the comments, I applied your change and pushed again, apologize for my mistakes last time.
> 
> Regarding the acc test import steps, I think we still need to keep the ignore step for those app setting:
> 
> ```
> "app_settings.WEBSITE_CONTENTSHARE", "app_settings.WEBSITE_CONTENTAZUREFILECONNECTIONSTRING", "app_settings.%"
> ```
> 
> As I mentioned in my previous comments, service will try to add the `WEBSITE_CONTENTSHARE` and `WEBSITE_CONTENTAZUREFILECONNECTIONSTRING` for the function app instance (in elastic or consumption plan) automatically if no `WEBSITE_CONTENTOVERVNET` is added by user, which means, we need to determine whether to set those kvp back to state by checking the user's TF config and that's why we need to ignore the changes when doing the import. So I'm not going to change the original unpackappsetting function:[`main`/internal/services/appservice/linux_function_app_resource.go#L1090](https://github.com/hashicorp/terraform-provider-azurerm/blob/main/internal/services/appservice/linux_function_app_resource.go?rgh-link-date=2023-03-15T02%3A34%3A14Z#L1090)

Hi @xiaxyi - Apologies, you are correct. These _should_ be ignored on Import as the service does not manage them if absent. I'll message you offline to discuss.

jackofallops commented:
Tests look good

<img width="654" alt="image" src="https://user-images.githubusercontent.com/11830746/225307238-ff0b6067-8452-4172-a6fa-ea5caea8194a.png">

jackofallops reviewed (approved):
Thanks @xiaxyi for the updates and patience. This LGTM now 👍
